### PR TITLE
feat(fetch): add support for file urls in fetch

### PR DESCRIPTION
**Pull request by Linkgoron**
This PR adds support for file urls in fetch

refs: https://github.com/nodejs/node/issues/42003

According to the fetch spec, file urls are "left as an exercise for the reader."
https://fetch.spec.whatwg.org/#scheme-fetch

### CODE REVIEW

File: lib/fetch/index.js [hunk 1]
```diff
@@ -882,7 +883,7 @@ async function schemeFetch (fetchParams) {
     case 'file:': {
       // For now, unfortunate as it is, file URLs are left as an exercise for the reader.
```
Comment (benjamingr) [inline lib/fetch/index.js:884]: remove comment?

File: lib/fetch/index.js [hunk 2]
```diff
@@ -1990,4 +1991,72 @@ function httpNetworkFetch (
   })
 }
 
+async function fileFetch (fetchParams) {
+  const { request } = fetchParams
+  const url = requestCurrentURL(request)
+  const context = this
```
Comment (benjamingr) [inline lib/fetch/index.js:1997]: why is context just not `this`?

File: lib/fetch/index.js [hunk 3]
```diff
@@ -1990,4 +1991,72 @@ function httpNetworkFetch (
...
+  if (request.method !== 'GET') {
+    return makeNetworkError(`Fetching files only supports the GET method. Received ${request.method}`)
```
Comment (benjamingr) [inline lib/fetch/index.js:1999]: What do browsers do with HEAD?

File: lib/fetch/index.js [hunk 4]
```diff
@@ -1990,4 +1991,72 @@ function httpNetworkFetch (
...
+  }
+
+  try {
+    function onTerminated () {
+      const aborted = context.terminated.aborted
+      if (aborted) {
+        response.aborted = true
+        stream.destroy(new AbortError())
+      } else {
+        stream.destroy(new TypeError('terminated'))
```
Comment (benjamingr) [inline lib/fetch/index.js:2035]: Not specific to these PR but these errors should have codes?

(That seems orthogonal though)

File: test/fetch/client-fetch.js [hunk 5]
```diff
@@ -332,3 +334,23 @@ test('post FormData with File', (t) => {
...
+  t.plan(2)
+  const fileToRead = pathToFileURL(join(__dirname, '..', 'fixtures', 'does-not-exist.text'))
+  try {
+    await fetch(fileToRead)
+    t.fail('fetch should have thrown')
+  } catch (err) {
+    t.ok(err.cause)
+    t.equal(err.cause.code, 'ENOENT')
+  }
+})
```
Comment (benjamingr) [inline test/fetch/client-fetch.js:356]: test AbortSignal?

**Review by benjamingr — COMMENTED** [review]
Good progress

File: lib/fetch/index.js (hunk 4 above)
Comment (Linkgoron) [reply lib/fetch/index.js:2035]: I looked at what's done in other methods.

File: lib/fetch/index.js (hunk 3 above)
Comment (Linkgoron) [reply lib/fetch/index.js:1999]: I looked at what Deno does, I event kept the message the same.

**Review by mcollina — CHANGES_REQUESTED** [review]
We need a way to control the security of this. There are quite a few possible attacks when reading files from disk on servers.

My opinion is that we should aim to a "close by default" approach vs "open by default".

I would recommend we add a "FileSystemEnabledAgent" that can direct requests to disk in case we need it to be. This entity will be responsible to configure what paths could be served.

**Comment by benjamingr** [comment]
> We need a way to control the security of this. There are quite a few possible attacks when reading files from disk on servers.

So your concern is that we get an external URL from the user and they send over a `file` url and that lets them somehow access/manipulate data on the server?

**Comment by mcollina** [comment]
> This kind of change would make this less portable (which was one of the reasons why this was requested), as I assume that other implementation don't have such constructions (I'll check them out). Is going out of spec and adding a flag or forcing localURLsOnly when using file urls an option?

That's not correct. In the browser this is managed by CSP and others. Unfortunately Node.js is not a browser and we have to deal with more strict security constraints, so I don't think this can be enabled by default without checks.

**Comment by mcollina** [comment]
> So your concern is that we get an external URL from the user and they send over a file url and that lets them somehow access/manipulate data on the server?

Exactly that. Moreover I might want to send a URL to read a file that is restricted in a given folder.

**Comment by Linkgoron** [comment]
> > This kind of change would make this less portable (which was one of the reasons why this was requested), as I assume that other implementation don't have such constructions (I'll check them out). Is going out of spec and adding a flag or forcing localURLsOnly when using file urls an option?
> 
> That's not correct. In the browser this is managed by CSP and others. Unfortunately Node.js is not a browser and we have to deal with more strict security constraints, so I don't think this can be enabled by default without checks.

Yes, of course. What I meant was that we I didn't think the fetch _method_ had such parameters. I edited my message after you already quoted me, where I asked if you meant that we should add some kind of global configuration for fetch that would allow such behaviour?

**Comment by mcollina** [comment]
We usually include those configuration inside the `Dispatcher` system. That's the right place where the retrieving of file should happen and where the checks should be.

**Comment by benjamingr** [comment]
Asked about this in Deno btw: https://github.com/denoland/deno/issues/11925#issuecomment-1043217870

**Comment by KhafraDev** [comment]
I think there should also be some warning that this behavior can change at any times; once (or if - file uris have been left to the implementation for years now) the spec includes steps for handling file:// urls, this will have to be rewritten and it's very likely that the behavior will change.

I believe that only Firefox and Deno handle file urls at the moment as well, which means it's not technically spec compliant and not widely available.

**Comment by Jamesernator** [comment]
While it's not a web option, could we just have an option to `fetch` to allow file urls? Browsers would simply ignore such an option so it would still be possible to write a compatible subset of `fetch` that works in both for loading package resources:

```js
// file.js
const myResource = new URL("./resource.txt", import.meta.url);
 
const resourceBlob = await fetch(myResource, {
    // If file.js is loaded in Node.js then myResource will be some file:// url in which
    // case Node.js will see this option and allow it,
    // In browsers myResource will generally be a https:// url and will just ignore this
    // options proceeding per normal
    allowFileURLs: true,
});
```

**Comment by Linkgoron** [comment]
> While it's not a web option, could we just have an option to `fetch` to allow file urls? Browsers would simply ignore such an option so it would still be possible to write a compatible subset of `fetch` that works in both for loading package resources:
> 
> ```js
> // file.js
> const myResource = new URL("./resource.txt", import.meta.url);
>  
> const resourceText = await fetch(myResource, {
>     // If file.js is loaded in Node.js then myResource will be some file:// url in which
>     // case Node.js will see this option and allow it,
>     // In browsers myResource will generally be a https:// url and will just ignore this
>     // options proceeding per normal
>     allowFileURLs: true,
> }).then(res => res.text());
> ```
> 
> By having such an option, other `fetch` calls could be presumed to be "safe" (to the extent arbitrary fetches are actually safe in node).

I agree that this feature is useful, and I'd be happy to get this into the undici if we can agree on how to make everyone happy with its security.

**Comment by mcollina** [comment]
Go ahead and add this option, I think it's a good addition. I would use a more ugly-and-prefixed name for it to avoid any potential conflicts with whatwg. Anyhow, I'd recommend you to open an issue at wintercg/fetch.

**Comment by KhafraDev** [comment]
I would prefer to leave this to https://github.com/wintercg/fetch/issues/4 instead of implementing a non-spec compliant solution.
